### ISSUE TITLE
[BUG]  Abnormal loss in deepspeed v0.17.2 + ulysess training, not decreasing.

### ISSUE THREAD

**Issue by kechunFIVE** [microsoft/DeepSpeed#7473]
the code is based on https://github.com/deepspeedai/DeepSpeed/blob/master/docs/_tutorials/ulysses-alst-sequence-parallelism.md but training with a wiki.train.raw dataset

```
# train.py
from deepspeed.runtime.sequence_parallel.ulysses_sp import UlyssesSPAttentionHF, UlyssesSPDataLoaderAdapter
from deepspeed.runtime.utils import move_to_device
from deepspeed.utils import groups
from torch import tensor
from transformers import AutoModelForCausalLM
import deepspeed
import deepspeed.comm as dist
import torch

model_name_or_path = '/data01/kechun.wu/models/Qwen2___5-7B-Instruct'
max_length = 2048
sequence_parallel_size = 4
micro_batch_size = 2

config_dict = {
    "train_micro_batch_size_per_gpu": 1,
    "zero_optimization": {
        "stage": 2,
    },
    "optimizer": {
        "type": "Adam",
        "params": {
            "lr": 1e-3
        }
    },
    "sequence_parallel_size": sequence_parallel_size,
}

zero_opt_dict = {
        "stage": 1,
        "stage3_param_persistence_threshold": 1e4,
        "stage3_max_live_parameters": 3e7,
        "memory_efficient_linear": False,
        #"reduce_scatter": True,
        "contiguous_gradients": True,
        "reduce_bucket_size": 1e9,
        "overlap_comm": False,
        #"stage3_prefetch_bucket_size": 5e8
    }

config_dict = {
    "train_micro_batch_size_per_gpu": micro_batch_size ,
    "steps_per_print": 1,
    "zero_optimization": zero_opt_dict,
    "optimizer": {
        "type": "Adam",
        "params": {
            "lr": 1e-5
        }
    },
    
    "fp16": {
        "enabled": False,
        "loss_scale_window": 100
    },
    "bf16": {
        "enabled": True
    },
    "gradient_clipping": 1.0,
    "sequence_parallel_size": sequence_parallel_size,
    }

dtype = torch.bfloat16

import os
import torch
from torch.utils.data import Dataset, DataLoader
from transformers import AutoTokenizer

class SimplePretrainingDataset(Dataset):
    def __init__(self, tokenizer, max_length=1024):

        self.tokenizer = tokenizer
        self.max_length = max_length
        self.tokenizer.pad_token = tokenizer.eos_token 
        ls = open('wiki.train.raw','r').readlines()
        self.texts = [t for t in ls if len(t) > 256]
    
    def __len__(self):
        return len(self.texts)
    
    def __getitem__(self, idx):
        text = self.texts[idx]
       
        tokenized = self.tokenizer(
            text,
            padding="max_length",
            truncation=True,
            max_length=self.max_length,
            return_tensors="pt"
        )
        
        input_ids = tokenized["input_ids"].squeeze(0)  
        attention_mask = tokenized["attention_mask"].squeeze(0)
        
        position_ids = torch.arange(0, len(input_ids), dtype=torch.long)
        labels = input_ids.clone()
      
        return {
            "input_ids": input_ids,
            "attention_mask": attention_mask,
            "position_ids": position_ids,
            "labels": labels
        }




dist.init_distributed(dist_backend='nccl', dist_init_required=True)

# Ulysses injection into HF Transformers
mpu = UlyssesSPAttentionHF.register_with_transformers(
    model_name_or_path=model_name_or_path,
    core_attn_implementation="sdpa",
    sequence_parallel_size=sequence_parallel_size,
    max_length=max_length,
    micro_batch_size=micro_batch_size,
    seq_length_is_variable=True,
)

# Deepspeed setup
model = AutoModelForCausalLM.from_pretrained(model_name_or_path)
tokenizer = AutoTokenizer.from_pretrained(model_name_or_path)

model, _, _, _ = deepspeed.initialize(config=config_dict,
                                        model=model,
                                        model_parameters=model.parameters(),
                                        mpu=mpu)

# UlyssesSPDataLoaderAdapter injection
sp_group = groups._get_sequence_parallel_group()
sp_world_size = groups._get_sequence_parallel_world_size()
sp_rank = groups._get_sequence_parallel_rank()

dataset = SimplePretrainingDataset(tokenizer, max_length=max_length)

dl = DataLoader(
    dataset,
    batch_size=micro_batch_size,
    shuffle=True,
    num_workers=0 
)
dl = UlyssesSPDataLoaderAdapter(
    dl,
    sp_rank=sp_rank,
    sp_group=sp_group,
    sp_world_size=sp_world_size,
    device=model.device,
)



for iter, batch in enumerate(dl):
    batch = move_to_device(batch, model.device)
    outputs = model(**batch)
    # as of this writing HF doesn't calculate loss with shift_labels yet and requires us to do it manually (liger does that automatically)
    shift_labels = batch["shift_labels"]
    #print( model.module.loss_function)
    loss = model.module.loss_function(
        logits=outputs.logits.cpu(),
        labels=None,
        shift_labels=shift_labels.cpu(),
        vocab_size=model.module.config.vocab_size,
    )

    loss = loss.to(outputs.logits.device)
    # differentiable weighted per-shard-loss aggregation across ranks
    losses_per_rank = torch.distributed.nn.functional.all_gather(loss, group=sp_group)
    # special dealing with SFT that has prompt tokens that aren't used in loss computation
    good_tokens = sum((shift_labels != -100).view(-1))
    good_tokens_per_rank = torch.distributed.nn.functional.all_gather(good_tokens, group=sp_group)
    total_loss = sum(losses_per_rank[rank] * good_tokens_per_rank[rank] for rank in range(sp_world_size))
    total_good_tokens = sum(good_tokens_per_rank)
    loss = total_loss / total_good_tokens

    model.backward(loss)
    
    if dist.get_rank() == 0:
        print(f"{iter}: {loss.item()=}")
```

below is the output, which i think is abnormal 

```
0: loss.item()=14.492203712463379
1: loss.item()=12.210344314575195
2: loss.item()=14.868824005126953
3: loss.item()=14.9736967086792
4: loss.item()=13.96434211730957
5: loss.item()=14.097953796386719
6: loss.item()=14.70736026763916
7: loss.item()=14.421201705932617
8: loss.item()=14.07357406616211
9: loss.item()=14.057064056396484
10: loss.item()=14.76125717163086
11: loss.item()=13.963306427001953
12: loss.item()=14.375652313232422
13: loss.item()=13.730012893676758
14: loss.item()=14.553333282470703
15: loss.item()=14.587777137756348
16: loss.item()=13.890111923217773
17: loss.item()=14.763873100280762
18: loss.item()=13.909299850463867
19: loss.item()=14.886909484863281
20: loss.item()=14.904806137084961
21: loss.item()=14.2420940399169921
```
